experimental/examples/ordered-list/src/orderedlist.ts

```diff
@@ -0,0 +1,166 @@
+    }
+
+    private move(destination: StablePlace, source: StableRange) {
+        this.tree.applyEdit(...Move.create(source, destination));
+    }
+
+    private insert(place: StablePlace, item: Serializable<T>): Position {
+        const newNode = makePositionNode(item);
+        this.tree.applyEdit(...Insert.create([newNode], place));
+        return newNode.identifier;
+    }
+
+    /** Inserts the given `item` at the beginning of the list. */
+    public insertFirst(item: Serializable<T>): Position {
+        return this.insert(this.firstPlace, item);
+    }
+
+    /** Inserts the given `item` at the end of the list. */
+    public insertLast(item: Serializable<T>): Position {
+        return this.insert(this.lastPlace, item);
+    }
+
+    /** Inserts the given `item` before the item at `nextSibling`. */
+    public insertBefore(nextSibling: Position, item: Serializable<T>): Position {
+        return this.insert(this.placeBefore(nextSibling), item);
+    }
+
+    /** Inserts the given `item` after the item at `prevSibling`. */
+    public insertAfter(prevSibling: Position, item: Serializable<T>): Position {
+        return this.insert(this.placeAfter(prevSibling), item);
+    }
+
+    /** Moves the item at `position` to the beginning of the list. */
+    public moveFirst(position: Position) {
+        this.move(this.firstPlace, StableRange.only(position));
+    }
+
+    /** Moves the item at `position` to the end of the list. */
+    public moveLast(position: Position)  {
+        this.move(this.lastPlace,  StableRange.only(position));
+    }
+
+    /** Moves the item at `position` before the item at `nextSibling`. */
+    public moveBefore(nextSibling: Position, position: Position) {
+        this.move(this.placeBefore(nextSibling), StableRange.only(position));
+    }
+
+    /** Moves the item at `position` after the item at `prevSibling`. */
+    public moveAfter(prevSibling: Position, position: Position) {
+        this.move(this.placeAfter(prevSibling), StableRange.only(position));
+    }
+
+    /** Removes the item at `position`. */
+    public remove(position: Position) {
+        this.tree.applyEdit(Delete.create(StableRange.only(position)));
+    }
+
+    /** Returns the item at `position`. */
+    public get(position: Position): Serializable<T> {
+        return this.view.getSnapshotNode(position).payload as Serializable<T>;
+    }
+
+    /** Sets the item at `position` to the given `value`. */
+    public set(position: Position, value: Serializable<T>) {
+        this.view.setNodeValue(position, value);
+    }
+
+    /** Removes all items from the list. */
+    public clear() {
+        this.tree.applyEdit(
+            Delete.create(
+                StableRange.all({
+                    parent: this.view.root,
+                    label: itemsTrait,
+                })));
+    }
+
+    /** Returns an array containing the positions of all items in the list. */
+    public get positions() {
```

Comment:
Curious, why OrderedList doesn't provide any iterator over elements but an array of positions ?
Also, if this is the way to get elements: 

```ts
list.positions.map((position) => list.get(position)); 
```
- Are position and _item at position_ guaranteed to remain the same when we do this?
- Does it return snapshot of positions  ?  (positions might have been added or removed when we call getPosition but will not be included in results?)

Reply:
That is the correct way to get the array of items.

Fluid schedules processes batches of ops in microtasks that run interleaved with your code, but never at the same time as your code (i.e., there's no shared memory or parallel threads.)

Therefore, the DDS will only change while you're awaiting a Promise or between asynchronous callbacks/events, but never during a synchronous block of code.  Therefore, the above code snippet is safe.

There's no reason OrderedList couldn't build the list of items and support an iterator, but there's also no advantage to having the OrderedList do this vs. you building what MFS needs on top.  (e.g., I suspect you'll want to keep items associated with their positions...  like '{ position: Position, item: T }[]').